homeassistant/components/qbus/coordinator.py

```diff
@@ -0,0 +1,274 @@
+
+        while self._cleanup_callbacks:
+            cleanup_callback = self._cleanup_callbacks.pop()
+            cleanup_callback()
+
+        self._cleanup_callbacks = []
+        self._controller_activated = False
+        self._subscribed_to_controller_state = False
+        self._controller = None
+
+    async def async_update_controller_config(self, config: QbusDiscovery) -> None:
+        """Update the controller based on the config."""
+        _LOGGER.debug("%s - Updating config", self._entry.unique_id)
+        serial = self._entry.data.get(CONF_SERIAL_NUMBER, "")
+        controller = config.get_device_by_serial(serial)
+
+        if controller is None:
+            _LOGGER.warning(
+                "%s - Controller with serial %s not found",
+                self._entry.unique_id,
+                serial,
+            )
+            return
+
+        self._controller = controller
+
+        self._update_device_info(self._controller)
+        await self._async_subscribe_to_controller_state(self._controller)
+        await self.async_refresh()
+        self._request_controller_state(self._controller)
+        self._request_entity_states(self._controller)
+
+    def _update_device_info(self, controller: QbusMqttDevice) -> None:
+        device_registry = dr.async_get(self.hass)
+        device_registry.async_get_or_create(
+            config_entry_id=self._entry.entry_id,
+            identifiers={(DOMAIN, format_mac(controller.mac))},
+            manufacturer=MANUFACTURER,
+            model="CTD3.x",
+            name=f"CTD {controller.serial_number}",
+            serial_number=controller.serial_number,
+            sw_version=controller.version,
+        )
+
+    async def _async_subscribe_to_controller_state(
+        self, controller: QbusMqttDevice
+    ) -> None:
+        if self._subscribed_to_controller_state is True:
+            return
+
+        controller_state_topic = self._topic_factory.get_device_state_topic(
+            controller.id
+        )
+        _LOGGER.debug(
+            "%s - Subscribing to %s",
+            self._entry.unique_id,
+            controller_state_topic,
+        )
+        self._subscribed_to_controller_state = True
+        self._cleanup_callbacks.append(
+            await mqtt.async_subscribe(
+                self.hass,
+                controller_state_topic,
+                self._controller_state_received,
+            )
+        )
+
+    async def _controller_state_received(self, msg: ReceiveMessage) -> None:
+        _LOGGER.debug(
+            "%s - Receiving controller state %s", self._entry.unique_id, msg.topic
+        )
+
+        if self._controller_activated:
+            return
+
+        state = self._message_factory.parse_device_state(msg.payload)
+
+        if (
+            state
+            and state.properties
+            and state.properties.connectable is False
+            and self._controller is not None
```

Comment:
can we move the check for `self._controller is None` to the top, before parsing the payload? it is probably never happening, but since it is a "sanity" check we can do it right away. it also makes this check a bit simpler